Userland/tee.cpp

```diff
@@ -0,0 +1,96 @@
+#include <AK/Vector.h>
+#include <fcntl.h>
+#include <getopt.h>
+#include <signal.h>
+#include <stdio.h>
+#include <unistd.h>
+
+static Vector<int> collect_fds(int argc, char** argv, int start, bool aflag, bool* err)
+{
+    int oflag;
+    mode_t mode;
+    if (aflag) {
+        oflag = O_APPEND;
+        mode = 0;
+    } else {
+        oflag = O_CREAT | O_WRONLY | O_TRUNC;
+        mode = S_IROTH | S_IWOTH | S_IRGRP | S_IWGRP | S_IRUSR | S_IWUSR;
+    }
+
+    Vector<int> fds;
+    for (int i = start; i < argc; ++i) {
+        int fd = open(argv[i], oflag, mode);
+        if (fd < 0) {
+            perror("failed to open file for writing");
+            *err = true;
+        }
+        fds.append(fd);
+    }
+    fds.append(STDOUT_FILENO);
+    return fds;
+}
+
+static void copy_stdin(Vector<int>& fds, bool* err)
+{
+    for (;;) {
+        char buf[4096];
+        ssize_t nread = read(STDIN_FILENO, buf, sizeof(buf));
+        if (nread == 0)
+            break;
+        if (nread < 0) {
+            perror("read() error");
+            *err = true;
+        }
+
+        for (int fd : fds) {
+            ssize_t nwrite = write(fd, buf, nread);
+            if (nwrite != nread)
```

Comment:
It's okay for `write()` to do a partial write, we shouldn't consider that an error, but instead retry until we've written everything. `write()` may also fail with `EINTR`, in which case I think we should also retry.

For other error codes, maybe we should print an error message to `stderr`?
And shouldn't we return on error?

Reply:
Yes, sorry! partial writes should be allowed. `EINTR` will have to be handled, yes. The other error codes should have an error message, I agree. According to my understanding of the docs, we should stop trying to write to this current file, yes, but we should continue writing the other successfully opened files.